yellow-paper/docs/cryptography/performance-targets.md

```diff
@@ -0,0 +1,121 @@
+* Time to compute a 2x2 rollup proof
+* Memory required to compute a 2x2 rollup proof
+
+<!-- We can break these properties down into metrics linked to specitic cryptographic components:
+
+* Size of Goblin Plonk proofs
+* Size of Honk proofs
+* Honk prover time
+* Goblin Plonk prover time
+* Protogalaxy recursion -->
+
+ "MVP" = minimum standards that we can go to main-net with.
+
+| metric | how to measure | MVP | ideal |
+| --- | --- | --- | --- |
+| proof size | total size of a user tx incl. goblin plonk proofs | 32kb | 8kb |
+| prover time | 8 iterations of protogalaxy with 2^17 circuits (web browser) | 1 min | 10 seconds |
+| verifier time | how long does it take the verifier to check a proof (incl. grumpkin IPA MSMs) | 25ms | 1ms |
+| memory consumption | fold 2^17 circuits into an accumulator an arbitrary number of times | 2gb | 512mb |
+| size of the kernel circuit | number of gates | 2^17 | 2^15 |
+| Aztec Virtual Machine prover time | 1 million VM step circuit | 30 seconds | 5 seconds |
+| Aztec Virtual Machine memory consumption | 1 million VM step circuit | 128Gb | 16Gb |
+
+### Proof size
+
+At a tx throughput of 1,024 tx per second, each Aztec node (not sequencer/prover, just a regular node that is sending transactions) needs to download `1024*proof_size` bytes of data to keep track of the mempool. 32kb proofs = 32MB per second which is on the threshold of practical outside of a datacenter. 8MB per second is much more practical.
+
+At launch, the throughput won't be this high, but ideally we architect Aztec such that the network can scale effectively by throwing hardware+resources at bottlenecks (easy to do in a post-decentralisation), without protocol, architecture or tech upgrades ( harder post-decentralisation).
+
+### Prover time
+
+The critical UX factor. 1 minute per user tx is on the cusp of workable. The faster we can get this the better our network's value proposition for users and developers
+
+### Verifier time
+
+This matters because verifying a transaction is effectively free work being performed by sequencers and network nodes that propagate txns to the mempool. If verification time becomes too large it opens up potential DDOS attacks.
+
+At a 1,024 transactions per second and 20ms verification time, a block builder needs to run 20 machines in parallel just to verify natively all of the proofs in a block! Not ideal but probably tolerable given this problem has a "more hardware go brrrr" solution.
+
+### Memory consumption
+
+This is *critical*. Users can tolerate slow proofs, but if Honk consumes too much memory, a user cannot make a proof at all.
+
+safari on iPhone will purge tabs that consume more than 1gb of RAM. 2gb will work in web browsers and some phones (but not all). 512mb will allow most phones to make proofs.
+
+### Kernel circuit size
+
+Not a critical metric, but the prover time + prover memory metrics are predicated on a kernel circuit costing about 2^17 constraints!
```

Comment:
I guess there are storage considerations, for resource-constrained devices?
A user might need to store the proving keys for various sizes of kernel circuit, and possibly commonly-used app circuits. Alternatively, they need to generate these proving keys on-the-fly, but we might run into memory & timing constraints, there?

During a tx, a user will need to store / hold in memory:
- Their DB of note/nullifier/hash path data.
- Newly created notes / nullifiers.
- Data to send alongside a proof, including data bus data (e.g. logs, enqueued public function args and data).
- Public inputs (and witnesses?) of functions which haven't-yet been verified within a kernel circuit iteration (or which haven't-yet been folded - I need to read more about folding).
- Witness & proving key & other polynomial stuff for the circuit currently being proven.
- ACIR for all functions that need to be executed in a tx.

Reply:
we don't need to store (in persistent storage) proving keys as they can be derived from Noir very quickly (if this is not currently the case, we need to make it so. The old cost of making proving keys was the required Fourier transforms, which do not exist for Honk).

I'm making an assumption that, without the proving keys, the persistent storage required for everything else is quite small.